feat: DAYL-123 Update Select List a11y docn (#5063)
BrightspaceUI/core: components/inputs/docs/input-select-styles.md

## Applying styles to native select elements

> Native `<select>` elements can be styled by importing `input-select-styles.js` into your LitElement and applying the `d2l-input-select` CSS class. The styles account for several CSS pseudo-classes, including `disabled`, `focus` and `hover`.

Reviewer: ```suggestion
Native `<select>` elements can be styled by importing `input-select-styles.js` into your LitElement and applying the `d2l-input-select` CSS class. 

The styles support the pseudo-classes `disabled`, `focus`, and `hover`, as well as the `aria-invalid` attribute.
```

<!-- docs: demo code name:d2l-test-input-select -->
```html
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';

  class TestInputSelect extends LitElement {
  static get styles() {
    return [ selectStyles ];
  }

  render() {
    return html`
      <select
        aria-label="Choose a dinosaur:"
        class="d2l-input-select">
        <option>Tyrannosaurus</option>
        <option>Velociraptor</option>
        <option>Deinonychus</option>
      </select>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select></d2l-test-input-select>
```

> ## Building a Custom Select Component

Reviewer: Hmm I feel like `RtlMixin` and `SkeletonMixin` should have separate documentation and perhaps just get called out here as something to be considered and then with a link to the docs.
Since they aren't in Daylight officially yet, you could point at their GitHub docs [here](https://github.com/BrightspaceUI/core/tree/main/components/skeleton#skeletizing-custom-elements-with-skeletonmixin) and [here](https://github.com/BrightspaceUI/core/tree/main/mixins/rtl#rtlmixin)

Author: So both those sections do link to their GitHub docs. I'm removing the `RtlMixin` section based on Dave's comment, but was planning on leaving the `SkeletonMixin` section. I figure it's still worth keeping and showing off what it looks like, while still linking to the doc, or would you prefer no code showing up?

Author: After talking in the meeting, I'll be removing the section and switching to just making a reference to the `SkeletonMixin` earlier on.

In addition to the `d2l-input-select` styling and [native attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#attributes), there are a few suggested Mixins and properties that we recommend using when creating a custom `Select` component.

### Invalid

Through the use of the [`aria-invalid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid) attribute, you are able to mark the input as invalid, which shows up both in the styling along with being picked up by screenreaders.

> <!-- docs: demo name:d2l-test-input-select -->

Reviewer: I think `name` is only needed if you want the properties table.
If we want these demos hooked up to the code sandbox, then they'll need unique `sandboxName` (Name of sandbox page if name is not set - unique identifier used for the url) and `sandboxTitle` (Human readable title for the sandbox page)

Reviewer 2: Similar to my comment for RTL, I think we could be more direct here, something like this:

> Use the [`aria-invalid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid) attribute to support screen reader users and apply our consistent error styles.

Should we also indicate how to wire up an error tooltip and ensure that's in the demo?

Author: Since they're just showing things off, just to show what it looks like, it probably doesn't need to be openable in sandbox, and if there should be one that's openable, then it would be the first one that would have several things in it.

```html
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';

  class TestInputSelect extends LitElement {

    static get properties() {
    return {
      invalid: { type: Boolean }
    };
  }

  static get styles() {
    return [ selectStyles ];
  }

  render() {
    const invalid = this.invalid ? 'true' : 'false';
    return html`
      <select
        aria-label="Choose a dinosaur:"
        aria-invalid="${invalid}"
        class="d2l-input-select">
        <option>Tyrannosaurus</option>
        <option>Velociraptor</option>
        <option>Deinonychus</option>
      </select>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select invalid></d2l-test-input-select>
```

### RtlMixin

> The [`RtlMixin`](https://github.com/BrightspaceUI/core/tree/main/mixins/rtl) is ideal for supporting right-to-left languages.

Reviewer 2: Maybe we should be more direct and say 

> Use the [`RtlMixin`](https://github.com/BrightspaceUI/core/tree/main/mixins/rtl) to support languages that read from right to left.

Reviewer 3: Actually, we [shouldn't be recommending new usages of RtlMixin](https://blog.d2l.dev/2024/05/31/state-of-the-ui/#right-to-forget-rtlmixin) at all, in favour of CSS logical properties.

Author: Got it, I'll remove the section about it then!

```

### SkeletonMixin

Using the [`SkeletonMixin`](https://github.com/BrightspaceUI/core/tree/main/components/skeleton) can help convey to users that the page, or at least a section of it, has not finished loading yet.

<!-- docs: demo name:d2l-test-input-select -->
```html
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';
  import { SkeletonMixin } from '@brightspace-ui/core/components/skeleton/skeleton-mixin.js';

  class TestInputSelect extends SkeletonMixin(LitElement) {

  static get styles() {
    return [ super.styles, selectStyles ];
  }

  render() {
    return html`
      <div class="d2l-skeletize">
        <select
          aria-label="Choose a dinosaur:"
          class="d2l-input-select">
          <option>Tyrannosaurus</option>
          <option>Velociraptor</option>
          <option>Deinonychus</option>
        </select>
      </div>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select skeleton></d2l-test-input-select>
```

## Accessibility

> - Due to merely being a CSS class, the accessibility provided by `d2l-input-select` comes purely in the way of following the guidelines for [contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and [focus](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html)

Reviewer: ```suggestion
- Due to merely being a CSS class, the accessibility provided by `selectStyles` comes purely in the way of following the guidelines for [contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and [focus](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html)
```

- There are several things that can be done to make sure your `select` component is accessible, including:
  - Following the W3C [Combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/) pattern
  - Using either the `aria-label` or `aria-labelledby` to appropriately assign a label to your component

>   - Also using `label` for `optgroup` if you choose to use that element within

Reviewer: ```suggestion
  - Using `label` for `optgroup` if you choose to use that element within
```
also maybe more clarify on within what

          <option>Velociraptor</option>
          <option>Deinonychus</option>
        </select>
      </div>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select></d2l-test-input-select>
```

### Invalid

Use the [`aria-invalid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid) attribute to support screenreader users and apply our consistent error styles.

<!-- docs: demo -->
```html
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';

  class TestInputSelect extends LitElement {

  static get styles() {
    return [ selectStyles ];
  }

  render() {
    return html`
      <select
        aria-label="Choose a dinosaur:"
        aria-invalid="true"
        class="d2l-input-select">
        <option>Tyrannosaurus</option>
        <option>Velociraptor</option>
        <option>Deinonychus</option>
      </select>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select></d2l-test-input-select>
```

## Accessibility

> - Due to merely being a CSS class, the accessibility provided by `selectStyles` comes purely in the way of following the guidelines for [contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and [focus](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html)

Reviewer: ```suggestion
- Due to applying styles based on a CSS class rather than being its own component, the accessibility provided by `selectStyles` comes purely in the way of following the guidelines for [contrast](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html) and [focus](https://www.w3.org/WAI/WCAG21/Understanding/focus-visible.html)
```

Native `<select>` elements can be styled by importing `input-select-styles.js` into your LitElement and applying the `d2l-input-select` CSS class.

The styles support the pseudo-classes `disabled`, `focus`, and `hover`, as well as the `aria-invalid` attribute.

When applying styles to the native element, we also recommend using the [`SkeletonMixin`](https://github.com/BrightspaceUI/core/tree/main/components/skeleton) to help convey to users that the page, or at least a section of it, has not finished loading yet.

> <!-- docs: demo code name:d2l-test-input-select -->

Reviewer: ```suggestion
<!-- docs: demo code -->
```

```html
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';
  import { SkeletonMixin } from '@brightspace-ui/core/components/skeleton/skeleton-mixin.js';

  class TestInputSelect extends SkeletonMixin(LitElement) {
  static get styles() {
    return [ super.styles, selectStyles ];
  }

  render() {
    return html`
      <div class="d2l-skeletize">
        <select
          aria-label="Choose a dinosaur:"
          class="d2l-input-select">
          <option>Tyrannosaurus</option>
          <option>Velociraptor</option>
          <option>Deinonychus</option>
        </select>
      </div>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select></d2l-test-input-select>
```

### Invalid

Use the [`aria-invalid`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid) attribute to support screenreader users and apply our consistent error styles.

<!-- docs: demo -->
```html

> <script type="module">
>   import { html, LitElement } from 'lit';
>   import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';
> 
>   class TestInputSelect extends LitElement {
> 
>     static get properties() {
>     return {
>       invalid: { type: Boolean }
>     };
>   }
> 
>   static get styles() {
>     return [ selectStyles ];
>   }
> 
>   render() {
>     const invalid = this.invalid ? 'true' : 'false';
>     return html`
>       <select
>         aria-label="Choose a dinosaur:"
>         aria-invalid="${invalid}"
>         class="d2l-input-select">
>         <option>Tyrannosaurus</option>
>         <option>Velociraptor</option>
>         <option>Deinonychus</option>
>       </select>
>     `;
>   }
> 
>   }
>   customElements.define('d2l-test-input-select', TestInputSelect);
> </script>
> <d2l-test-input-select invalid></d2l-test-input-select>

Reviewer: ```suggestion
<script type="module">
  import { html, LitElement } from 'lit';
  import { selectStyles } from '@brightspace-ui/core/components/inputs/input-select-styles.js';

  class TestInputSelect extends LitElement {

  static get styles() {
    return [ selectStyles ];
  }

  render() {
    return html`
      <select
        aria-label="Choose a dinosaur:"
        aria-invalid
        class="d2l-input-select">
        <option>Tyrannosaurus</option>
        <option>Velociraptor</option>
        <option>Deinonychus</option>
      </select>
    `;
  }

  }
  customElements.define('d2l-test-input-select', TestInputSelect);
</script>
<d2l-test-input-select></d2l-test-input-select>
```
Since this isn't actually showing the code you can simplify it.

Author: Had to make sure that was using `aria-invalid="true"`, since without the `="true"` it doesn't show up, which has always been something I've found weird, but I guess it's just a native element thing 🤷
